ipalib/rpc.py

```diff
@@ -999,77 +999,77 @@ def create_connection(self, ccache=None, verbose=None, fallback=None,
-                # We don't care about the response, just that we got one
-                break
-            except errors.KerberosError:
-                # kerberos error on one server is likely on all
-                raise
-            except ProtocolError as e:
-                if hasattr(context, 'session_cookie') and e.errcode == 401:
-                    # Unauthorized. Remove the session and try again.
-                    delattr(context, 'session_cookie')
+                    command = getattr(serverproxy, 'ping')
                     try:
-                        delete_persistent_client_session_data(principal)
-                    except Exception:
-                        # This shouldn't happen if we have a session but it isn't fatal.
-                        pass
-                    return self.create_connection(
-                        ccache, verbose, fallback, delegate)
-                if not fallback:
+                        command([], {})
+                    except Fault as e:
+                        e = decode_fault(e)
+                        if e.faultCode in errors_by_code:
+                            error = errors_by_code[e.faultCode]
+                            raise error(message=e.faultString)
+                        else:
+                            raise UnknownError(
+                                code=e.faultCode,
+                                error=e.faultString,
+                                server=url,
+                            )
+                    # We don't care about the response, just that we got one
+                    return serverproxy
+                except errors.KerberosError:
+                    # kerberos error on one server is likely on all
                     raise
-                else:
-                    self.log.info('Connection to %s failed with %s', url, e)
-                serverproxy = None
-            except Exception as e:
-                if not fallback:
-                    raise
-                else:
-                    self.log.info('Connection to %s failed with %s', url, e)
-                serverproxy = None
-
-        if serverproxy is None:
-            raise NetworkError(uri=_('any of the configured servers'),
-                error=', '.join(urls))
-        return serverproxy
+                except ProtocolError as e:
+                    if hasattr(context, 'session_cookie') and e.errcode == 401:
+                        # Unauthorized. Remove the session and try again.
+                        delattr(context, 'session_cookie')
+                        try:
+                            delete_persistent_client_session_data(principal)
+                        except Exception:
+                            # This shouldn't happen if we have a session but
+                            # it isn't fatal.
+                            pass
+                        # try once more with a new session
+                        break
```

Comment:
This break is related to the loop on urls. This means that when reached, we exit from the loop on urls and retry with the first url. This is not optimal if the first server is down as it will be retried many times. I would rather put the loops in the reverse order, i.e. when the session_cookie needs to be deleted, retry with the same url.